File: src/rs/cache/store/SectorChainStore.ts
import { ByteSource } from "../../io/ByteSource";
import { ByteSourceSlice } from "../../io/ByteSourceSlice";
import { CacheIndex } from "../CacheIndex";
import { CacheStore } from "./CacheStore";
import { Sector } from "./Sector";
import { SectorCluster } from "./SectorCluster";
import { readI32BE, readU16BE, readU24BE } from "../../io/Endian";

class SectorChainArchiveSource implements ByteSource {
    constructor(
        private readonly dataFile: ByteSource,
        private readonly sectorIds: number[],
        readonly size: number,
        private readonly headerSize: number,
        private readonly dataSize: number,
    ) {}

    slice(start: number, size: number): ByteSource {
        return new ByteSourceSlice(this, start, size);
    }

    tryGetUint8ArrayView(): Uint8Array | null {
        return null;
    }

    readInto(
        offset: number,
        target: Uint8Array,
        targetOffset: number = 0,
        length: number = target.length - targetOffset,
    ): void {
        if (length < 0) {
            throw new Error("Invalid length");
        }
        if (offset < 0 || offset + length > this.size) {
            throw new Error(`Read out of bounds. offset=${offset}, length=${length}, size=${this.size}`);
        }
        if (length === 0) {
            return;
        }

        let remaining = length;
        let inOff = offset;
        let outOff = targetOffset;

        while (remaining > 0) {
            const sectorIndex = (inOff / this.dataSize) | 0;
            const sectorOffset = inOff - sectorIndex * this.dataSize;
            const take = Math.min(this.dataSize - sectorOffset, remaining);

            const sectorId = this.sectorIds[sectorIndex];
            if (sectorId === undefined) {
                throw new Error(`Invalid sector index: ${sectorIndex}`);
            }

            const fileOffset = sectorId * Sector.SIZE + this.headerSize + sectorOffset;
            this.dataFile.readInto(fileOffset, target, outOff, take);

            remaining -= take;
            inOff += take;
            outOff += take;
        }
    }
}

export class SectorChainStore implements CacheStore {
    constructor(
        readonly dataFile: ByteSource,
        readonly indexFiles: (ByteSource | null)[],
        readonly metaFile: ByteSource | null,
    ) {}

    getIndexFileSize(indexId: number): number | null {
        const file = this.getIndexFile(indexId);
        return file ? file.size : null;
    }

    openArchiveReader(indexId: number, archiveId: number): ByteSource {
        if (indexId < 0) {
            throw new Error("Index id cannot be lower than 0");
        }

        const indexFile = this.getIndexFile(indexId);
        if (!indexFile) {
            throw new Error(`Index ${indexId} not found`);
        }

        const sectorIndexId = this.getSectorIndexId(indexId);
        const cluster = this.readSectorCluster(indexFile, indexId, archiveId);
        const size = cluster.size;
        const extended = archiveId > 65535;

        const headerSize = extended ? Sector.EXTENDED_HEADER_SIZE : Sector.HEADER_SIZE;
        const dataSize = extended ? Sector.EXTENDED_DATA_SIZE : Sector.DATA_SIZE;

        const sectorIds = this.walkSectorChain(
            sectorIndexId,
            archiveId,
            cluster.sector,
            size,
            extended,
        );

        return new SectorChainArchiveSource(
            this.dataFile,
            sectorIds,
            size,
            headerSize,
            dataSize,
        );
    }

    private getIndexFile(indexId: number): ByteSource | null {
        if (indexId === CacheIndex.META_INDEX_ID) {
            return this.metaFile;
        }
        return this.indexFiles[indexId] ?? null;
    }

    private getSectorIndexId(indexId: number): number {
        if (this.metaFile) {
            return indexId;
        }
        return indexId + 1;
    }

    private readSectorCluster(indexFile: ByteSource, indexId: number, archiveId: number): SectorCluster {
        const clusterPtr = archiveId * SectorCluster.SIZE;
        const fileSize = indexFile.size;
        if (clusterPtr < 0 || clusterPtr + SectorCluster.SIZE > fileSize) {
            throw new Error(
                `Invalid ptr: ${clusterPtr}, fileSize: ${fileSize}, indexId: ${indexId}, archiveId: ${archiveId}`,
            );
        }

        const buf = new Uint8Array(SectorCluster.SIZE);
        indexFile.readInto(clusterPtr, buf);

        const size = readU24BE(buf, 0);
        const sector = readU24BE(buf, 3);
        return new SectorCluster(size, sector);
    }

    private walkSectorChain(
        sectorIndexId: number,
        archiveId: number,
        firstSectorId: number,
        totalSize: number,
        extended: boolean,
    ): number[] {
        const headerSize = extended ? Sector.EXTENDED_HEADER_SIZE : Sector.HEADER_SIZE;
        const dataSize = extended ? Sector.EXTENDED_DATA_SIZE : Sector.DATA_SIZE;

        const header = new Uint8Array(headerSize);

        const sectorIds: number[] = [];
        let remaining = totalSize;
        let chunk = 0;
        let sectorId = firstSectorId;

        while (remaining > 0) {
            const sectorPtr = sectorId * Sector.SIZE;
            this.dataFile.readInto(sectorPtr, header);

            let readArchiveId: number;
            let readChunk: number;
            let nextSector: number;
            let readIndexId: number;

            if (extended) {
                readArchiveId = readI32BE(header, 0) >>> 0;
                readChunk = readU16BE(header, 4);
                nextSector = readU24BE(header, 6);
                readIndexId = header[9];
            } else {
                readArchiveId = readU16BE(header, 0);
                readChunk = readU16BE(header, 2);
                nextSector = readU24BE(header, 4);
                readIndexId = header[7];
            }

            if ((readArchiveId >>> 0) !== (archiveId >>> 0)) {
                throw new Error(`Sector archive id mismatch. expected: ${archiveId} got: ${readArchiveId}`);
            }
            if (readIndexId !== sectorIndexId) {
                throw new Error(`Sector index id mismatch. expected: ${sectorIndexId} got: ${readIndexId}`);
            }
            if (readChunk !== chunk) {
                throw new Error(`Sector chunk mismatch. expected: ${chunk} got: ${readChunk}`);
            }

            sectorIds.push(sectorId);

            chunk++;
            sectorId = nextSector;
            remaining -= dataSize;
        }

        return sectorIds;
    }

    // Note: forward chunk iteration can be implemented on top of `openArchiveReader` if/when needed.
}

File: cpp/rs/cache/store/SectorChainStore.cpp
#include "SectorChainStore.hpp"

#include <algorithm>
#include <cstddef>
#include <memory>
#include <optional>
#include <span>
#include <stdexcept>
#include <utility>
#include <vector>

#include "../../io/ByteSource.hpp"
#include "../../io/ByteSourceSlice.hpp"
#include "../../io/Endian.hpp"
#include "../../util/U32.hpp"
#include "../../types.hpp"
#include "DatLayout.hpp"

namespace rs {

class SectorChainArchiveSource final : public ByteSource, public std::enable_shared_from_this<SectorChainArchiveSource> {
public:
    SectorChainArchiveSource(
        ByteSourcePtr dataFile,
        std::vector<u32> sectorIds,
        std::size_t size,
        std::size_t headerSize,
        std::size_t dataSize)
        : dataFile_(std::move(dataFile)),
          sectorIds_(std::move(sectorIds)),
          size_(size),
          headerSize_(headerSize),
          dataSize_(dataSize) {}

    [[nodiscard]] std::size_t size() const override { return size_; }

    [[nodiscard]] ByteSourcePtr slice(std::size_t start, std::size_t size) const override {
        return std::make_shared<ByteSourceSlice>(shared_from_this(), start, size);
    }

    void readInto(std::size_t offset, u8* target, std::size_t length) const override {
        if (offset > size_ || length > size_ - offset) {
            throw std::out_of_range("SectorChainArchiveSource: read out of bounds");
        }
        if (length == 0) {
            return;
        }

        std::size_t remaining = length;
        std::size_t inOff = offset;
        std::size_t outOff = 0;

        while (remaining > 0) {
            const std::size_t sectorIndex = inOff / dataSize_;
            const std::size_t sectorOffset = inOff - sectorIndex * dataSize_;
            const std::size_t take = std::min<std::size_t>(dataSize_ - sectorOffset, remaining);

            if (sectorIndex >= sectorIds_.size()) {
                throw std::runtime_error("SectorChainArchiveSource: invalid sector index");
            }
            const u32 sectorId = sectorIds_[sectorIndex];

            const std::size_t fileOffset = static_cast<std::size_t>(sectorId) * SECTOR_SIZE + headerSize_ + sectorOffset;
            dataFile_->readInto(fileOffset, target + outOff, take);

            remaining -= take;
            inOff += take;
            outOff += take;
        }
    }

    [[nodiscard]] std::optional<std::span<const u8>> tryGetUint8ArrayView() const override { return std::nullopt; }

private:
    ByteSourcePtr dataFile_;
    std::vector<u32> sectorIds_;
    std::size_t size_;
    std::size_t headerSize_;
    std::size_t dataSize_;
};

SectorChainStore::SectorChainStore(
    ByteSourcePtr dataFile,
    std::vector<std::optional<ByteSourcePtr>> indexFiles,
    std::optional<ByteSourcePtr> metaFile)
    : dataFile_(std::move(dataFile)), indexFiles_(std::move(indexFiles)), metaFile_(std::move(metaFile)) {
    if (!dataFile_) {
        throw std::invalid_argument("SectorChainStore: dataFile is null");
    }
}

const ByteSource* SectorChainStore::getIndexFile(i32 indexId) const {
    if (indexId < 0) {
        return nullptr;
    }
    if (indexId == 255) {
        return metaFile_ ? metaFile_->get() : nullptr;
    }
    const std::size_t idx = static_cast<std::size_t>(indexId);
    if (idx >= indexFiles_.size()) {
        return nullptr;
    }
    const auto& opt = indexFiles_[idx];
    return opt ? opt->get() : nullptr;
}

i32 SectorChainStore::getSectorIndexId(i32 indexId) const {
    if (metaFile_) {
        return indexId;
    }
    return indexId + 1;
}

std::optional<std::size_t> SectorChainStore::getIndexFileSize(i32 indexId) const {
    const ByteSource* file = getIndexFile(indexId);
    if (!file) {
        return std::nullopt;
    }
    return file->size();
}

SectorChainStore::SectorCluster SectorChainStore::readSectorCluster(
    const ByteSource& indexFile,
    i32 indexId,
    i32 archiveId) const {
    (void)indexId;
    if (archiveId < 0) {
        throw std::out_of_range("SectorChainStore: archiveId < 0");
    }

    const std::size_t clusterPtr = static_cast<std::size_t>(archiveId) * IDX_ENTRY_SIZE;
    if (clusterPtr + IDX_ENTRY_SIZE > indexFile.size()) {
        throw std::out_of_range("SectorChainStore: idx entry out of bounds");
    }

    u8 buf[IDX_ENTRY_SIZE];
    indexFile.readInto(clusterPtr, buf, IDX_ENTRY_SIZE);

    SectorCluster c;
    c.size = readU24BE(buf);
    c.sector = readU24BE(buf + 3);
    return c;
}

std::vector<u32> SectorChainStore::walkSectorChain(
    i32 sectorIndexId,
    u32 archiveId,
    u32 firstSectorId,
    u32 totalSize,
    bool extended) const {
    const std::size_t headerSize = extended ? SECTOR_EXTENDED_HEADER_SIZE : SECTOR_HEADER_SIZE;
    const std::size_t dataSize = extended ? SECTOR_EXTENDED_DATA_SIZE : SECTOR_DATA_SIZE;

    std::vector<u32> sectorIds;
    u32 remaining = totalSize;
    u32 chunk = 0;
    u32 sectorId = firstSectorId;

    std::vector<u8> header;
    header.resize(headerSize);

    while (remaining > 0) {
        const std::size_t sectorPtr = static_cast<std::size_t>(sectorId) * SECTOR_SIZE;
        if (sectorPtr + headerSize > dataFile_->size()) {
            throw std::out_of_range("SectorChainStore: sector ptr out of bounds");
        }
        dataFile_->readInto(sectorPtr, header.data(), headerSize);

        u32 readArchiveId = 0;
        u32 readChunk = 0;
        u32 nextSector = 0;
        u32 readIndexId = 0;

        if (extended) {
            readArchiveId = toU32(readI32BE(header.data()));
            readChunk = readU16BE(header.data() + 4);
            nextSector = readU24BE(header.data() + 6);
            readIndexId = header[9];
        } else {
            readArchiveId = readU16BE(header.data());
            readChunk = readU16BE(header.data() + 2);
            nextSector = readU24BE(header.data() + 4);
            readIndexId = header[7];
        }

        if (toU32(readArchiveId) != toU32(archiveId)) {
            throw std::runtime_error("SectorChainStore: sector archive id mismatch");
        }
        if (readIndexId != static_cast<u32>(sectorIndexId)) {
            throw std::runtime_error("SectorChainStore: sector index id mismatch");
        }
        if (readChunk != chunk) {
            throw std::runtime_error("SectorChainStore: sector chunk mismatch");
        }

        sectorIds.push_back(sectorId);
        chunk++;
        sectorId = nextSector;
        if (remaining > dataSize) {
            remaining -= static_cast<u32>(dataSize);
        } else {
            remaining = 0;
        }
    }

    return sectorIds;
}

ByteSourcePtr SectorChainStore::openArchiveReader(i32 indexId, i32 archiveId) const {
    if (indexId < 0) {
        throw std::runtime_error("SectorChainStore: indexId < 0");
    }

    const ByteSource* indexFile = getIndexFile(indexId);
    if (!indexFile) {
        throw std::runtime_error("SectorChainStore: index file not found");
    }

    const i32 sectorIndexId = getSectorIndexId(indexId);
    const SectorCluster cluster = readSectorCluster(*indexFile, indexId, archiveId);
    const std::size_t size = static_cast<std::size_t>(cluster.size);
    const bool extended = archiveId > 65535;

    const std::size_t headerSize = extended ? SECTOR_EXTENDED_HEADER_SIZE : SECTOR_HEADER_SIZE;
    const std::size_t dataSize = extended ? SECTOR_EXTENDED_DATA_SIZE : SECTOR_DATA_SIZE;

    const std::vector<u32> sectorIds =
        walkSectorChain(sectorIndexId, static_cast<u32>(archiveId), cluster.sector, cluster.size, extended);

    return std::make_shared<SectorChainArchiveSource>(dataFile_, sectorIds, size, headerSize, dataSize);
}

} // namespace rs
